src/Usb.Net/Windows/WindowsUsbInterface.cs

```diff
@@ -71,29 +71,59 @@ public void Dispose()
             GC.SuppressFinalize(this);
         }
 
-        public Task<ReadResult> SendControlTransferAsync(SetupPacket setupPacket, byte[] inputBuffer, CancellationToken cancellationToken = default)
+        public Task<ControlTransferResult> SendControlTransferAsync(SetupPacket setupPacket, byte[] buffer, CancellationToken cancellationToken = default)
         {
             return Task.Run(() =>
             {
-                uint bytesWritten = 0;
+                var transferBuffer = Array.Empty<byte>();
 
-                if (setupPacket.Length > 0 && inputBuffer == null)
+                uint bytesTransferred = 0;
+
+                if (setupPacket.Length > 0)
                 {
-                    inputBuffer = new byte[setupPacket.Length];
+                    // the request has data exchange
+                    // which direction?
+                    if (setupPacket.RequestType.Direction == RequestDirection.In)
+                    {
+                        // data is to be received
+                        // create buffer of the appropriate size
+                        transferBuffer = new byte[setupPacket.Length];
+                    }
+                    else
+                    {
+                        // data is to be sent
+                        // TODO
+                        transferBuffer = new byte[SetupPacket.SetupPacketSize + setupPacket.Length];
```

Comment:
@josesimoes why not move the array.copy code here?

Reply:
Still haven't tested this fully, but I think the buffer has to be done concatenating the setup packet and the data buffer.

Reply:
@josesimoes OK yes. We definitely need a test for this before we move forward.